plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/semantics/SQLEditorSemanticMarkersManager.java

```diff
@@ -132,25 +129,34 @@ private void updateMarkers() {
         for (Map.Entry<SQLDocumentScriptItemSyntaxContext, Boolean> entry : entries) {
             SQLDocumentScriptItemSyntaxContext scriptItem = entry.getKey();
             if (entry.getValue() && scriptItem.getProblems() != null) {
+                Map<Integer, SQLSemanticErrorAnnotation> severestAnnotationsByLine = new HashMap<>();
                 Deque<SQLSemanticErrorAnnotation> itemAnnotations = this.annotations.computeIfAbsent(scriptItem, c -> new LinkedList<>());
                 int scriptItemPosition = scriptItem.getInitialPosition();
                 for (SQLQueryRecognitionProblemInfo problemInfo : scriptItem.getProblems()) {
                     try {
                         Interval problemInterval = problemInfo.getInterval();
+                        int problemOffset = scriptItemPosition + problemInterval.a;
+                        int problemLine = this.editor.getDocument().getLineOfOffset(problemOffset);
                         final IMarker marker = resource.createMarker(SQLSemanticErrorAnnotation.MARKER_TYPE, Map.of(
                                 IMarker.SEVERITY, problemInfo.getSeverity().markerSeverity,
                                 IMarker.MESSAGE, problemInfo.getMessage(),
                                 IMarker.TRANSIENT, true
                         ));
                         SQLSemanticErrorAnnotation annotation = new SQLSemanticErrorAnnotation(marker, problemInfo);
                         marker.setAttribute(SQLSemanticErrorAnnotation.MARKER_ATTRIBUTE_NAME, annotation);
-                        Position position = new Position(scriptItemPosition + problemInterval.a, problemInterval.length());
+                        Position position = new Position(problemOffset, problemInterval.length());
                         annotationModel.addAnnotation(annotation, position);
                         itemAnnotations.addLast(annotation);
-                    } catch (CoreException e) {
+                        severestAnnotationsByLine.compute(problemLine, (k, v) -> (v == null ||
+                            annotation.getProblemInfo().getSeverity().markerSeverity > v.getProblemInfo().getSeverity().markerSeverity
+                        ) ? annotation : v);
```

Comment:
This code is quite hard to read... :(
- You can use `Map#merge` that is identical to `Map#compute` but handles `null` keys automatically
- You can add getter for `getProblemInfo().getSeverity().markerSeverity` to shorten the code